### PR TITLE
Fixes #22742 - Add link delay option to subnet

### PR DESCRIPTION
**Pull request by jcpunk**
This commit adds a link delay option to a subnet.

This is primarily useful on links where there can be a few delays in getting connectivity.

### CODE REVIEW

**Comment by ares** [comment]
ok to test

from quick read it looks good, how is that used? Is there linked PR in community templates?

**Comment by jcpunk** [comment]
Once this is merged I'll add the bits to the community templates.  I didn't want to get too far ahead of myself.

**Comment by ares** [comment]
ah I see, it would be great to have both PRs ready so I could test it together :-) also it would be good if it lands in the same release

**Comment by jcpunk** [comment]
Template PR now exists

**Comment by jcpunk** [comment]
Just checking in

**Comment by mmoll** [comment]
@jcpunk please fix the rubocop violation

File: test/models/nic_test.rb [hunk 1]
```diff
@@ -92,6 +92,7 @@ def teardown
     assert_equal subnet.network, interface.network
     assert_equal subnet6.network, interface.network6
     assert_equal subnet.vlanid, interface.vlanid
+    assert_equal 3 subnet.nic_delay,
```
[the remark this answers is not in the document]
Comment (jcpunk) [reply test/models/nic_test.rb]: removed comma

File: app/models/nic/interface.rb [hunk 2]
```diff
@@ -61,6 +61,14 @@ def mtu
...
     end
 
+    def nic_delay
+      if self.attrs['nic_delay'].present? && self.attrs['nic_delay'] > 0
+        return self.attrs['nic_delay']
+      end
+      nil
+    end
+
+
```
[the remark this answers is not in the document]
Comment (jcpunk) [reply app/models/nic/interface.rb]: removed blank line

**Comment by jcpunk** [comment]
Should be fixed now

**Comment by mmoll** [comment]
@jcpunk Thanks, could you please also rebase to pull in the new Rubocop rules that went in  the last days?

**Comment by jcpunk** [comment]
rebased

File: test/models/subnet_test.rb [hunk 3]
```diff
@@ -145,6 +146,9 @@ class Subnet::Test < Subnet; end
     assert_equal 1500, Subnet.new.mtu
   end
 
+  test "should have nic_delay set to nil by default" do
+    assert_equal nil, Subnet.new.nic_delay
+
```
Comment (mmoll) [inline test/models/subnet_test.rb:152]: `end`? :)

Comment (jcpunk) [reply test/models/subnet_test.rb:152]: yark, that will teach me to do two things at once...

**Comment by mmoll** [comment]
test fails :broken_heart:

**Comment by jcpunk** [comment]
Hmmm.... perhaps my test is not testing rationally for the ruby world....

Does it make sense to be testing this?  From a workflow perspective nic_delay is like MTU.  If it isn't set you can ignore it an use the defaults.

But I feel weird not having a default and testing it....

**Comment by mmoll** [comment]
@jcpunk Please have a look at https://github.com/mmoll/foreman/commit/fe986b974f662ea2c7f2afceb29a0aa24609e364 which should get the tests green.

Maybe I did get something wrong, but shouldn't this be an optional setting, so null/nil is allowed? MTu is a bit different, as it 1500 by default and thus can always be set.

**Comment by jcpunk** [comment]
I'll take a look at that code.

On the null front, to my mind `nil` and `0` are the same in this instance as zero is the default behavior of "no delay".

**Comment by jcpunk** [comment]
I think I've successfully cloned your changes into this PR.

**Comment by mmoll** [comment]
@jcpunk You can of course also have the number 0 as default, just some of the surroundings and the test needs to get updated. I don't really know, what's better here. Having the number 0 or having nil is a difference here, a bit like in mathematics between 0 and an empty set.

**Comment by jcpunk** [comment]
For now it is probably fine to mirror what is done with MTU.  I supposed if we need to convert `nil` to `0` down the road it is easy to do... the reverse not so much.

**Comment by jcpunk** [comment]
rebased per request.

File: app/models/nic/interface.rb [hunk 4]
```diff
@@ -69,6 +69,10 @@ def bridge
       attrs[:bridge]
     end
 
+    def nic_delay
+      return self.attrs['nic_delay'] if self.attrs['nic_delay'].present?
```
Comment (tbrisker) [inline app/models/nic/interface.rb]: Should this be delegated to the subnet?
I don't see anywhere where we set this attribute on the interface or use it?

**Review by tbrisker — CHANGES_REQUESTED** [review]
Test failures are unrelated and were fixed in https://github.com/theforeman/foreman/commit/2fd66c014de4cf2b8b0057cad2f1c7901588eae3. 
Thanks @jcpunk and sorry for the long delay in getting this in! 
Just one small question inline, other than that I believe this is ready to merge.

File: app/models/nic/interface.rb (hunk 4 above)
Comment (jcpunk) [reply app/models/nic/interface.rb]: You're right, that should be from the subnet.  I can drop this as it doesn't make sense...

**Review by tbrisker — APPROVED** [review]
Thanks @jcpunk and sorry again for the slow turnaround on this one! Will merge once tests finish running.

**Pull request merged**